### PR TITLE
Make xray object table credis-managed and hence flushable.

### PR DESCRIPTION
**Pull request by concretevitamin**
This is a followup to https://github.com/ray-project/ray/pull/2266 -- that PR laid down the flushing policy framework + made task table flushable.  This PR in addition makes xray's task table flushable.

Testing program: submitting N noops in a loop, monitor memory usage of the _redis data shard_.  Built and ran with both `RAY_USE_NEW_GCS` and `RAY_USE_XRAY`.
```python
import subprocess
import time

import redis

import ray

# Pass this flag to avoid annoying output on program exit.
info = ray.init(redirect_worker_output=True)

# Turn on flushing with the next two lines.
pol = ray.experimental.SimpleGcsFlushPolicy(
    flush_when_at_least_bytes=0,
    flush_period_secs=1e-4,
    flush_num_entries_each_time=10000)
ray.experimental.set_flushing_policy(pol)

primary_port = info['redis_address'].split(':')[-1]
primary = redis.StrictRedis('127.0.0.1', primary_port)
splits = primary.lrange('RedisShards', 0, -1)[0].split(b':')
shard_port = splits[-1].decode()
print('ports: primary {} shard {}'.format(primary_port, shard_port))

measure_shard = True

port = primary_port
if measure_shard:
    port = shard_port

pid = subprocess.run(
    ['pgrep', '-f', 'redis-server.*{}'.format(port)],
    stdout=subprocess.PIPE).stdout[:-1].decode()
logfile = 'psrecord-{}.log'.format(port)


@ray.remote
def noop():
    return


time.sleep(1)
args = [
    'psrecord', pid, '--interval', '1', '--duration', '60', '--log', logfile
    # 'psrecord', pid, '--interval', '1', '--duration', '100000', '--log', logfile
]
print(' '.join(args))
subprocess.Popen(args)

# Assuming each noop generates ~3 writes to the GCS, we need a flush throughput
# of ~2.1k/s to break even.  <- Use this a sanity check.
num_noops = 200000
i = 0
start = time.time()
while i < num_noops:
    noop.remote()
    i += 1
print('Time to execute {} no-ops: {:.2f} seconds'.format(
    num_noops,
    time.time() - start))
```

### CODE REVIEW

File: python/ray/monitor.py [hunk 1]
```diff
@@ -111,10 +111,14 @@ def __init__(self, redis_address, redis_port, autoscaling_config):
         self.issue_gcs_flushes = "RAY_USE_NEW_GCS" in os.environ
         self.gcs_flush_policy = None
         if self.issue_gcs_flushes:
-            # For now, we take the primary redis server to issue flushes,
-            # because task table entries are stored there under this flag.
+            # Data is stored under the first data shard, so we issue flushes to
```
Comment (pcmoritz) [inline python/ray/monitor.py]: Let's add a note or todo here that this needs to change when we implement sharding

File: python/ray/monitor.py [hunk 2]
```diff
@@ -111,10 +111,14 @@ def __init__(self, redis_address, redis_port, autoscaling_config):
...
+            # that redis server.
+            addr_port = self.redis.lrange("RedisShards", 0, -1)[0]
+            addr_port = addr_port.split(b":")
```
Comment (pcmoritz) [inline python/ray/monitor.py]: addr, port = addr_port.split(b":") would be slightly cleaner but up to you

File: python/ray/monitor.py [hunk 3]
```diff
@@ -111,10 +111,14 @@ def __init__(self, redis_address, redis_port, autoscaling_config):
...
+            self.redis_shard = redis.StrictRedis(
+                host=addr_port[0], port=addr_port[1])
             try:
-                self.redis.execute_command("HEAD.FLUSH 0")
+                self.redis_shard.execute_command("HEAD.FLUSH 0")
```
Comment (stephanie-wang) [inline python/ray/monitor.py]: Why was this change necessary? Is `self.redis` pointing to the wrong server?

Also, is this code run every time a flush is required? If so, we should try to avoid running the `self.redis.lrange` command every time.

File: src/ray/gcs/tables.cc [hunk 4]
```diff
@@ -19,8 +24,16 @@ Status Log<ID, Data>::Append(const JobID &job_id, const ID &id,
   flatbuffers::FlatBufferBuilder fbb;
   fbb.ForceDefaults(true);
   fbb.Finish(Data::Pack(fbb, dataT.get()));
-  return context_->RunAsync("RAY.TABLE_APPEND", id, fbb.GetBufferPointer(), fbb.GetSize(),
-                            prefix_, pubsub_channel_, std::move(callback));
+
+  std::string command;
```
Comment (pcmoritz) [inline src/ray/gcs/tables.cc]: We have this code at enough places now that it warrants a little helper function I think

File: src/ray/gcs/tables.cc [hunk 5]
```diff
@@ -42,8 +55,16 @@ Status Log<ID, Data>::AppendAt(const JobID &job_id, const ID &id,
   flatbuffers::FlatBufferBuilder fbb;
   fbb.ForceDefaults(true);
   fbb.Finish(Data::Pack(fbb, dataT.get()));
-  return context_->RunAsync("RAY.TABLE_APPEND", id, fbb.GetBufferPointer(), fbb.GetSize(),
-                            prefix_, pubsub_channel_, std::move(callback), log_length);
+
+  std::string command;
```
Comment (pcmoritz) [inline src/ray/gcs/tables.cc]: similar to the above

File: src/common/redis_module/ray_redis_module.cc [hunk 6]
```diff
@@ -717,23 +696,77 @@ int TableAppend_RedisCommand(RedisModuleCtx *ctx,
...
-      return RedisModule_ReplyWithSimpleString(ctx, "OK");
-    }
+
   } else {
     // The requested index did not match the current length of the log. Return
     // an error message as a string.
-    const char *reply = "ERR entry exists";
-    return RedisModule_ReplyWithStringBuffer(ctx, reply, strlen(reply));
+    static const char *reply = "ERR entry exists";
+    RedisModule_ReplyWithStringBuffer(ctx, reply, strlen(reply));
```
Comment (pcmoritz) [inline src/common/redis_module/ray_redis_module.cc]: can we use RedisModule_ReplyWithError here?

**Review by pcmoritz — COMMENTED** [review]
Looks good modulo small fixes (see comments)

File: src/common/redis_module/ray_redis_module.cc [hunk 7]
```diff
@@ -717,23 +696,77 @@ int TableAppend_RedisCommand(RedisModuleCtx *ctx,
...
-    // Publish a message on the requested pubsub channel if necessary.
-    TablePubsub pubsub_channel = ParseTablePubsub(pubsub_channel_str);
-    if (pubsub_channel != TablePubsub::NO_PUBLISH) {
-      // All other pubsub channels write the data back directly onto the
-      // channel.
-      return PublishTableAdd(ctx, pubsub_channel_str, id, data);
-    } else {
-      return RedisModule_ReplyWithSimpleString(ctx, "OK");
-    }
+
```
Comment (stephanie-wang) [inline src/common/redis_module/ray_redis_module.cc]: I think you need to reply with a Redis error using `RedisModule_ReplyWithError` here.

File: src/common/redis_module/ray_redis_module.cc [hunk 8]
```diff
@@ -717,23 +696,77 @@ int TableAppend_RedisCommand(RedisModuleCtx *ctx,
...
   } else {
     // The requested index did not match the current length of the log. Return
     // an error message as a string.
-    const char *reply = "ERR entry exists";
-    return RedisModule_ReplyWithStringBuffer(ctx, reply, strlen(reply));
+    static const char *reply = "ERR entry exists";
+    RedisModule_ReplyWithStringBuffer(ctx, reply, strlen(reply));
+    return REDISMODULE_ERR;
+  }
+  return REDISMODULE_OK;
```
Comment (stephanie-wang) [inline src/common/redis_module/ray_redis_module.cc]: Move this into the `if` block above? Since it's only reachable by that branch.

File: src/ray/gcs/tables.cc [hunk 9]
```diff
@@ -3,6 +3,11 @@
 #include "common_protocol.h"
 #include "ray/gcs/client.h"
 
+namespace {
+static const std::string kTableAppendCommand = "RAY.TABLE_APPEND";
```
Comment (stephanie-wang) [inline src/ray/gcs/tables.cc]: Should we also move the `TABLE_ADD` commands here?

**Review by stephanie-wang — COMMENTED** [review]
Left a few comments. Overall, looks great!

File: src/common/redis_module/ray_redis_module.cc (hunk 7 above)
Comment (concretevitamin) [reply src/common/redis_module/ray_redis_module.cc]: Where?  If you mean `    RAY_CHECK(flags == REDISMODULE_ZADD_ADDED) << "Appended a duplicate entry";` then I think no since (1) it will crash (2) the existing behavior is like so too?

File: src/common/redis_module/ray_redis_module.cc (hunk 8 above)
Comment (concretevitamin) [reply src/common/redis_module/ray_redis_module.cc]: Done, good call

File: src/ray/gcs/tables.cc (hunk 9 above)
Comment (concretevitamin) [reply src/ray/gcs/tables.cc]: Not really since that code is not duplicated.  Also I don't see a good way to abstract that piece across append and add, without doing costly / unwieldy string concatenation.  Here the strings are compile-time constants.

**Comment by concretevitamin** [comment]
Addressed comments, PTAL, thanks!

**Review by pcmoritz — APPROVED** [review]

File: src/ray/gcs/tables.cc (hunk 9 above)
Comment (stephanie-wang) [reply src/ray/gcs/tables.cc]: Hmm, I just meant to also define `static const std::string`s for the `TABLE_ADD` commands.

I'm fine with not trying to abstract it out right now, but in the future we could still define the strings statically while doing the concatenation in the definition. As we add more commands, it would be nice to have a static table of commands that we could lookup as part of the helper function that you added in this PR.

File: src/common/redis_module/ray_redis_module.cc [hunk 10]
```diff
@@ -717,23 +696,76 @@ int TableAppend_RedisCommand(RedisModuleCtx *ctx,
...
-    // Publish a message on the requested pubsub channel if necessary.
-    TablePubsub pubsub_channel = ParseTablePubsub(pubsub_channel_str);
-    if (pubsub_channel != TablePubsub::NO_PUBLISH) {
-      // All other pubsub channels write the data back directly onto the
-      // channel.
-      return PublishTableAdd(ctx, pubsub_channel_str, id, data);
-    } else {
-      return RedisModule_ReplyWithSimpleString(ctx, "OK");
-    }
+    return REDISMODULE_OK;
```
Comment (stephanie-wang) [inline src/common/redis_module/ray_redis_module.cc]: We may need to be more careful about which functions (node vs. tail) should return a Redis reply and when, to make sure that every Redis command returns exactly one reply. I think it should be:
1. If node func errors, then it should return an error reply.
2. If node func does not error, then it should not return anything.
3. Tail func should always return a Redis reply.

I think this means that for the master commit on [credis](https://github.com/ray-project/credis/blob/master/src/chain_module.h#L352), we are incorrectly handling the case where the node func does not error and the command is executed on a non-tail node. In this case, the node func does not return a reply and the tail func isn't run, so we won't be returning a Redis reply.

**Comment by concretevitamin** [comment]
Travis `LINUX_WHEELS=1` build kept failing with no obvious errors.  Mind taking a look @pcmoritz ?

File: src/common/redis_module/ray_redis_module.cc (hunk 10 above)
Comment (concretevitamin) [reply src/common/redis_module/ray_redis_module.cc]: @stephanie-wang yes, bullets 1-3 are exactly what I have in mind.

For singleton chain case (default deployment), both `ActAsTail()` and `ActAsHead()` will be true.  So if node func does not error, the tail func _will_ be run, and hence we will exactly get 1 reply.

For distributed, >1-chain case, the `ChainReplicate()` interface specifically designed for ray's redis module lacks this distributed support (i.e., propagating the update).  (credis' native `MemberPut_RedisCommand()/Put()` do handle propagation, of course.) . Since it's not handled right now we should make sure 1-node case is correct at the moment.

File: src/ray/gcs/tables.cc (hunk 9 above)
Comment (concretevitamin) [reply src/ray/gcs/tables.cc]: Done

**Comment by pcmoritz** [comment]
@concretevitamin This also happens on the current master so is unrelated to this PR. I'm investigating it.

File: src/common/redis_module/ray_redis_module.cc (hunk 10 above)
Comment (stephanie-wang) [reply src/common/redis_module/ray_redis_module.cc]: Sounds good, just something to keep in mind once we move to distributed.
